File: desktop/electron/src/ipc/migration.ts
/// Migration state read/write command family (ADR-055 M1) — the Electron
/// equivalent of `src-tauri/src/migration.rs`. Same command names
/// (`migration_read` / `migration_export`) and JSON contract, so
/// `src/migration/state.ts` drives them unchanged through the bridge.
///
/// The renderer's boot flow (`importStateIfFresh` before render, then
/// `armExport`) snapshots the migration-covered localStorage keys to
/// `<user-data>/migration/state-v1.json` and restores them into a fresh
/// Chromium profile. `migration_read` first looks at Electron's own `userData`
/// copy, then falls back to the **Tauri** install's app-data dir — the
/// one-time cross-install handoff (plan §5), pulled forward from M3 because it
/// is a read-only fallback and device-testable today (#353). The restore
/// imports into localStorage and the next `migration_export` writes Electron's
/// own copy, so the legacy path is consulted at most once.
import { app } from 'electron';
import os from 'node:os';
import path from 'node:path';
import { mkdir, readFile, rename, writeFile } from 'node:fs/promises';
import type { Handler } from './dispatch';

function stateDir(): string {
  return path.join(app.getPath('userData'), 'migration');
}
function statePath(): string {
  return path.join(stateDir(), 'state-v1.json');
}

/// Where the Tauri shell's egress (M0.2) writes, per platform — the Tauri
/// app-data dir keyed by the bundle identifier (`tauri.conf.json`). Mirrors
/// Tauri's `app_data_dir`: macOS `~/Library/Application Support/<id>`,
/// Windows `%APPDATA%\<id>`, Linux `${XDG_DATA_HOME:-~/.local/share}/<id>`.
/// (Electron's own `appData` would be wrong on Linux — it is XDG_CONFIG_HOME.)
function legacyStatePaths(): string[] {
  const id = 'app.termipod.desktop';
  if (process.platform === 'linux') {
    const dataHome = process.env.XDG_DATA_HOME ?? path.join(os.homedir(), '.local', 'share');
    return [path.join(dataHome, id, 'migration', 'state-v1.json')];
  }
  return [path.join(app.getPath('appData'), id, 'migration', 'state-v1.json')];
}

/// Crash-safe write: temp file + atomic rename, matching the Rust command.
async function writeAtomic(json: string): Promise<void> {
  const dir = stateDir();
  await mkdir(dir, { recursive: true });
  const finalPath = statePath();
  const tmp = `${finalPath}.tmp`;
  await writeFile(tmp, json, 'utf8');
  await rename(tmp, finalPath);
}

export const migrationHandlers: Record<string, Handler> = {
  migration_export: async (args: Record<string, unknown>) => {
    const json = typeof args.json === 'string' ? args.json : '';
    if (json === '') return;
    await writeAtomic(json);
  },

  migration_read: async (): Promise<string | null> => {
    for (const candidate of [statePath(), ...legacyStatePaths()]) {
      try {
        return await readFile(candidate, 'utf8');
      } catch {
        /* absent — try the next candidate */
      }
    }
    return null; // no snapshot anywhere — a genuine first boot
  },
};

File: desktop/src-tauri/src/migration.rs
//! Migration data egress (ADR-055 / plan M0).
//!
//! localStorage is bound to the webview profile and will NOT follow the app
//! into Electron's Chromium profile. So the frontend snapshots every
//! `termipod.*` key to a versioned JSON file under app-data; Electron's first
//! boot re-imports it (M1), and it doubles as a free local backup. Bytes never
//! leave the device.

use std::fs;
use std::path::PathBuf;

use tauri::{AppHandle, Manager};

/// `<app-data>/migration/state-v1.json`. The `migration/` dir is created on
/// demand (mirrors `storage::attachment_default_dir`).
fn state_path(app: &AppHandle) -> Result<PathBuf, String> {
    let base = app.path().app_data_dir().map_err(|e| e.to_string())?;
    let dir = base.join("migration");
    fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    Ok(dir.join("state-v1.json"))
}

/// Persist the frontend's serialized localStorage snapshot. Written via a temp
/// file + rename so a crash mid-write never leaves a truncated JSON file.
#[tauri::command]
pub async fn migration_export(app: AppHandle, json: String) -> Result<(), String> {
    let path = state_path(&app)?;
    let tmp = path.with_extension("tmp");
    fs::write(&tmp, json.as_bytes()).map_err(|e| e.to_string())?;
    fs::rename(&tmp, &path).map_err(|e| e.to_string())?;
    Ok(())
}

/// Read back the snapshot, or `Ok(None)` if none has been written yet.
#[tauri::command]
pub async fn migration_read(app: AppHandle) -> Result<Option<String>, String> {
    let path = state_path(&app)?;
    match fs::read_to_string(&path) {
        Ok(s) => Ok(Some(s)),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(None),
        Err(e) => Err(e.to_string()),
    }
}
